lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp

```diff
@@ -835,6 +836,176 @@ class RAIIOpenedArchetypesTracker {
   }
 };
 
+/// Recursively traverse the conformance lists to determine sole conforming
+/// class, struct or enum type.
+NominalTypeDecl *
+SILCombiner::findSoleConformingTypeFromPCA(ProtocolDecl *Protocol) {
+  SmallVector<ProtocolDecl *, 16> PDWorkList;
+  SmallPtrSet<ProtocolDecl *, 8> VisitedPDs;
+  SmallPtrSet<NominalTypeDecl *, 8> ConformanceSet;
+  PDWorkList.push_back(Protocol);
+  while (!PDWorkList.empty()) {
+    auto *PD = PDWorkList.pop_back_val();
+    // Protocols must have internal or lower access scopes.
+    if (PD->getEffectiveAccess() > AccessLevel::Internal) {
+      return nullptr;
+    }
+    VisitedPDs.insert(PD);
+    auto NTDList = PCA->getConformances(PD);
+    if (NTDList.empty())
+      continue;
+    for (auto *ConformingNTD : NTDList) {
+      // Sole conforming class should not be public or have derived classes.
+      ClassDecl *CD;
+      if ((CD = dyn_cast<ClassDecl>(ConformingNTD)) &&
+          (CD->getEffectiveAccess() > AccessLevel::Internal ||
+           CHA->hasKnownDirectSubclasses(CD) ||
+           CHA->hasKnownIndirectSubclasses(CD))) {
+        return nullptr;
+      }
+      // Recurse on protocol types.
+      if (auto *Proto = dyn_cast<ProtocolDecl>(ConformingNTD)) {
+        // Ignore visited protocol decls.
+        if (!VisitedPDs.count(Proto))
+          PDWorkList.push_back(Proto);
+      } else { // Classes, Structs and Enums are added here.
+        ConformanceSet.insert(ConformingNTD);
+      }
+    }
+    // Bail early if we already found two or more concrete conformances.
+    if (ConformanceSet.size() > 1) {
+      return nullptr;
+    }
+  }
+  // Final check for only one conforming type with internal or lower access.
+  if (ConformanceSet.size() != 1)
+    return nullptr;
+
+  return *(ConformanceSet.begin());
+}
+
+/// Helper routine for preparing and propagating concrete type from
+/// protocol conformance analysis.
+bool SILCombiner::prepareNCheckConcreteTypePropagationFromPCA(
+    FullApplySite Apply, ProtocolDecl *Protocol, SILValue &Arg,
+    SILValue &NewArg, Optional<ProtocolConformanceRef> &ConformanceRef,
+    ArchetypeType *OpenedArchetype, CanType &ConcreteType) {
+
+  /// Only applicable in whole-module compilation.
+  if (!Apply.getModule().isWholeModule())
+    return false;
+
+  /// Ignore already visited Apply sites.
+  if (VisitedAIForPCA.find(Apply) != VisitedAIForPCA.end())
+    return false;
+  VisitedAIForPCA.insert(Apply);
+
+  // Ensure the Access level of the protocol is internal or lower.
+  if (Protocol->getEffectiveAccess() > AccessLevel::Internal)
+    return false;
+
+  // Determine the sole conforming type.
+  auto *NTD = findSoleConformingTypeFromPCA(Protocol);
+  if (!NTD)
+    return false;
+
+  // Do not handle unbounded generics.
+  auto ElementType = NTD->getDeclaredType();
```

Comment:
Better way to check: NTD->isGenericContext(). And move this into findSoleConformingTypeFromPCA()